gflownet/utils/batch.py

```diff
@@ -836,22 +859,66 @@ def get_rewards(
     def _compute_rewards(self):
         """
         Computes rewards for all self.states by first converting the states into proxy
-        format.
-
-        Returns
-        -------
-        rewards: torch.tensor
-            Tensor of rewards.
+        format. The result is stored in self.rewards as a torch.tensor
         """
-        states_proxy_done = self.get_terminating_states(proxy=True)
+
         self.rewards = torch.zeros(len(self), dtype=self.float, device=self.device)
         done = self.get_done()
-        if len(done) > 0:
+        if self.non_terminal_rewards:
+            self.rewards = self.env.proxy2reward(self.env.proxy(self.states2proxy()))
+        elif len(done) > 0:
+            states_proxy_done = self.get_terminating_states(proxy=True)
             self.rewards[done] = self.env.proxy2reward(
                 self.env.proxy(states_proxy_done)
             )
         self.rewards_available = True
 
+    def get_rewards_parents(self) -> TensorType["n_states"]:
+        """
+        Returns the rewards of all parents in the batch
+        """
+        if not self.rewards_parents_available:
+            self._compute_rewards_parents()
+        return self.rewards_parents
+
+    def _compute_rewards_parents(self):
+        """
+        Computes rewards of the self.parents by reusing rewards of the states (i.e. self.rewards).
+        Stores the result in self.rewards_parents
+        """
+        state_rewards = self.get_rewards()
+        self.rewards_parents = torch.zeros_like(state_rewards)
+        parent_is_source = self.get_parent_is_source()
+        parent_indices = self.get_parents_indices()
```

Comment:
This is what I meant in https://github.com/alexhernandezgarcia/gflownet/pull/253/files#r1406451240:
```suggestion
        parent_indices = self.get_parents_indices()
        parent_is_source = parent_indices == -1
```